driver/src/main/java/org/neo4j/driver/internal/security/SSLContextManager.java

```diff
@@ -0,0 +1,141 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.neo4j.driver.internal.security;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.security.KeyException;
+import java.security.KeyManagementException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import org.neo4j.driver.ClientCertificateManager;
+import org.neo4j.driver.Logger;
+import org.neo4j.driver.Logging;
+import org.neo4j.driver.exceptions.ClientException;
+import org.neo4j.driver.internal.InternalClientCertificate;
+import org.neo4j.driver.internal.pki.PemParser;
+import org.neo4j.driver.internal.util.Futures;
+
+class SSLContextManager {
+    private final ClientCertificateManager clientCertificateManager;
+    private final SecurityPlan.SSLContextSupplier sslContextSupplier;
+    private final Logger logger;
+    private CompletableFuture<SSLContext> sslContextFuture;
+    private SSLContext sslContext;
+
+    public SSLContextManager(
+            ClientCertificateManager clientCertificateManager,
+            SecurityPlan.SSLContextSupplier sslContextSupplier,
+            Logging logging) {
+        this.clientCertificateManager = clientCertificateManager;
+        this.sslContextSupplier = sslContextSupplier;
+        logger = logging.getLog(getClass());
+    }
+
+    public CompletionStage<SSLContext> get() {
+        return clientCertificateManager != null
+                ? getSSLContextWithClientCertificate()
+                : getSSLContextWithoutClientCertificate();
+    }
+
+    private CompletionStage<SSLContext> getSSLContextWithClientCertificate() {
+        CompletableFuture<SSLContext> sslContextFuture;
+        CompletionStage<SSLContext> sslContextStage = null;
+        synchronized (this) {
+            if (this.sslContextFuture == null) {
+                this.sslContextFuture = new CompletableFuture<>();
+                sslContextFuture = this.sslContextFuture;
+                var sslContext = this.sslContext;
+                sslContextStage = clientCertificateManager.getCertificate().thenApply(clientCertificate -> {
+                    var certificate = (InternalClientCertificate) clientCertificate;
+                    if (certificate.hasUpdate() || sslContext == null) {
```

Comment:
Whe is hasUpdate set to be false?

Reply:
It is either set by the user in the `ClientCertificateManager` implementation or by the provided `RotatingClientCertificateManager` implementation.